### DISCUSSION TITLE
address(this) and msg.sender

### DISCUSSION THREAD

**Discussion by zhongguoliliang** [Cyfrin/foundry-full-course-cu#1595] [General]
Hai! I was confused by address(this) and msg.sender! Who can tell me  the difference between them? what situation use address (this) and what situation use msg.sender？Thank you!

**Comment by EngrPips** [comment]
Hello @zhongguoliliang. 
Address(this) : This is the address of the contract you are writing 

Msg.sender : This is the address that send a transaction to a contract

You use addres(this) when you want to refer to the contract that you are working with and you use msg.sender when you want to refer to the address that send transaction to the contract that you are working with
Did I make it clear enough?

**Reply by zhongguoliliang** [reply]
Yes!  Great!

**Reply by EngrPips** [reply]
Awesome, glad you get it!
